### PR TITLE
Use Transformer in-memory with stdin/stdout

### PR DESCRIPTION
**Pull request by pseeth**
This PR tries to address #6. I basically followed the code by @carlthome in pysndfx here: https://github.com/carlthome/python-audio-effects/blob/master/pysndfx/dsp.py#L472. I followed the discussion in the issue and implemented a new function that belongs to Transformer called `build_array`, which takes in a numpy array. The main issue is keeping track of all of the information that would normally be in the header of the audio file. Instead, these have to be passed as arguments to build_array, both for the input and output.

I modified all of the tests. Each piece of the Transformer is tested by taking an input file and passing it through sox to write to an output file. So in the tests, what I do is load both the input file and the output file as numpy arrays, pass the input array into `tfm.build_array` and collect the output array (which is the second argument in the tuple, matching the API of `tfm.build`). Then, I do an `np.allclose` between the loaded output array from the output file and the output array. Every test was modified in this way, except for one, which I can't get working in the time I tried to put this together today. That test is `test_bitdepth_valid`, here: https://github.com/pseeth/pysox/blob/master/tests/test_transform.py#L1347.

This change to the tests is encapsulated in a single function `tfm_assert_array_to_file_output` which you'll see calls to sprinkled throughout the test.

To do all this, I modified the `sox` function in a backwards compatible way. I probably need to up coverage a bit still but this is hopefully okay to PR now for feedback.

Let me know if this is a good start and how to get this merged, if possible! It would go a long way to making pysox super efficient for Scaper, thus why I'm here. :)

Thanks!

### CODE REVIEW

**Comment by hadware** [comment]
The import of numpy could probably be made optional (if it's deemed to be too "heavy" of a dependency for this package).

Great work though!

**Comment by pseeth** [comment]
I got this up to 100% coverage in `transforms.py` but I'm getting coverage has dropped for some reason still. As far as I can tell, all the lines I wrote are covered but there are some missing lines in core, as there was before. Let me know if I missed something...

I can easily make numpy optional by checking for it or importing it dynamically but I'm guessing everyone using pysox for something might already have numpy installed, anyway?

**Comment by rabitt** [comment]
@pseeth this is great, thank you! 🚀 I'm reviewing now.

File: sox/transform.py [hunk 1]
```diff
@@ -454,6 +461,102 @@ def build(self, input_filepath, output_filepath, extra_args=None,
...
+            Bitdepth of input audio. If None, inferred at runtime by sox. 
+        bits_out : int, default=None
+            Expected bitdepth of output audio. If None, inferred by sox.
+        encoding_out : str or None, default=None
+            str determining expected encoding of output audio. Must be one of
+            's8', 's16', 'f32', 'f64'.
+        extra_args : list or None, default=None
+            If a list is given, these additional arguments are passed to SoX
+            at the end of the list of effects.
+            Don't use this argument unless you know exactly what you're doing!
```
Comment (rabitt) [inline sox/transform.py:498]: Document the returns

File: sox/transform.py [hunk 2]
```diff
@@ -454,6 +461,102 @@ def build(self, input_filepath, output_filepath, extra_args=None,
...
+            channels_out = channels_in
+        
+        if sample_rate_out is None:
+            sample_rate_out = sample_rate_in
+
+        encoding = input_array.dtype.type
+        if encoding_out is None:
+            encoding_out = encoding
+
+        self.set_input_format(
```
Comment (rabitt) [inline sox/transform.py:517]: I'm not sure we want to call these functions inside build_array - they change the internal state of the Transformer. The way it's written now, calling `build_array` will silently change the behavior of `build`.

Brainstorming how to get around this, we could:
1. write a function that builds the args list for input/output formats, which gets called by `set_input_format`/`set_output_format` and by this function. 
2. reduce the number of inputs to this function. as far as I can tell, the only needed input format argument is `sample_rate_in`. The rest can (should!?) be inferred from the input array. The output format support is already built into sox.build in several ways, either by calling `set_output_format` or by transformer commands like `rate` and `channels`. The only one I'd leave optionally specified here is `encoding_out`.

so, in short this function would become
```
build_array(self, input_array, sample_rate_in, encoding_out=None, 
            extra_args=None,  return_output=True)
```

Thoughts/objections?

File: sox/transform.py [hunk 3]
```diff
@@ -454,6 +461,102 @@ def build(self, input_filepath, output_filepath, extra_args=None,
...
+            Expected bitdepth of output audio. If None, inferred by sox.
+        encoding_out : str or None, default=None
+            str determining expected encoding of output audio. Must be one of
+            's8', 's16', 'f32', 'f64'.
+        extra_args : list or None, default=None
+            If a list is given, these additional arguments are passed to SoX
+            at the end of the list of effects.
+            Don't use this argument unless you know exactly what you're doing!
+        '''
+        output_filepath = '-'
```
Comment (rabitt) [inline sox/transform.py:500]: Thinking out loud... does it also make sense to support array in --> file out, and file in --> array out? My thinking is yes, in which case maybe all or this function should be part of `build`? What do you think @pseeth ?

**Comment by rabitt** [comment]
by the way, I'm totally OK with adding numpy as a dependency.

File: sox/transform.py (hunk 2 above)
Comment (pseeth) [reply sox/transform.py:517]: So the argument list for `build_array` was initially what you said, but had to be expanded quite a bit to get through all of the test cases. The ones I can get rid of are `channels_in` and `bits_in` as they can be inferred from the numpy array. But the output types can't be inferred. 

What if we had `set_output_format` and `set_input_format` return (optionally) the arg list rather than set it as a variable in the object? Then we could extend it dynamically in `build_array` instead without side effects.

**Comment by pseeth** [comment]
I'm not sure if I can get this to work for the Combiner as there are multiple input arrays and I'm not sure how that works with stdin. Seems like some black magic...but if a user is using numpy arrays after the transformer, they really should just use `np.stack` or `sum` to implement functionality like the combiner as it's much faster than doing an exec call to `sox`.

File: sox/core.py [hunk 4]
```diff
@@ -37,15 +51,28 @@ def sox(args):
...
-            args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
+        if src_array is not None and isinstance(src_array, np.ndarray):
+            process_handle = subprocess.Popen(
+                args, 
+                stdin=subprocess.PIPE, 
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            )
+            out, err = process_handle.communicate(src_array.T.tobytes(order='F'))
```
[the remark this answers is not in the document]
Comment (pseeth) [reply sox/core.py:61]: So to be honest, this is copied from @carlthome's solution in `pysndfx`. Looking into it a bit, I think this only matters for multichannel audio, as the two channels get interleaved (I think?) in 'F', but not in 'C'. I think this is how `sox` expects the audio?

[the remark this answers is not in the document]
Comment (pseeth) [reply sox/core.py:61]: Flipping from 'C' to 'F' resulted in 7 failed tests so I believe that yeah 'F' is load bearing.

File: sox/transform.py (hunk 3 above)
Comment (pseeth) [reply sox/transform.py:500]: Totally possible. Do you have a suggested function signature for `build`? It seems like a lot of the arguments will become optional, or do type-checking on input/output (in which case the arguments will be renamed). Then we'll have some sort of merge between the `build_array` and `build` functions, which I think would look reasonable. Let me know your thoughts!

File: setup.py [hunk 5]
```diff
@@ -20,12 +20,14 @@
         keywords='audio effects SoX',
         license='BSD-3-Clause',
         install_requires=[
+            'numpy',
```
[the remark this answers is not in the document]
Comment (pseeth) [reply setup.py:23]: I _think_ it's 1.9.0. According to this: https://numpy.org/doc/1.18/reference/generated/numpy.ndarray.tobytes.html, it's new in 1.9.0.

File: setup.py [hunk 6]
```diff
@@ -20,12 +20,14 @@
...
         ],
         extras_require={
             'tests': [
                 'pytest',
                 'pytest-cov',
                 'pytest-pep8',
+                'pysoundfile',
```
[the remark this answers is not in the document]
Comment (pseeth) [reply setup.py:30]: This is just for tests, but I think we need at least 0.9.0 to get the `dtype` functionality:

https://pysoundfile.readthedocs.io/en/latest/#breaking-changes

File: setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pseeth) [reply setup.py]: Updated!

File: setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pseeth) [reply setup.py]: Updated!

File: sox/core.py [hunk 7]
```diff
@@ -37,15 +51,28 @@ def sox(args):
     try:
         logger.info("Executing: %s", ' '.join(args))
 
-        process_handle = subprocess.Popen(
-            args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
+        if src_array is not None and isinstance(src_array, np.ndarray):
```
[the remark this answers is not in the document]
Comment (pseeth) [reply sox/core.py]: Yeah makes sense! I updated this.

File: sox/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pseeth) [reply sox/core.py:74]: Added a comment above this line.

File: sox/transform.py (hunk 2 above)
Comment (pseeth) [reply sox/transform.py]: I modified `set_[input, output]_format` with a flag `return_only` where the arg list will be built without saving it to `self`. I had to keep some of the arguments to ` build_array`, but as you'll see further down, `build_array` is now merged with `build`.

File: sox/transform.py (hunk 3 above)
Comment (pseeth) [reply sox/transform.py]: This is supported now! I merged `build` and `build_array` to be able to support this. There's some finicky typechecking that needed to happen, and a flag that tells the `core.sox` function whether to decode `out` with `utf-8` but it now works! I expanded the test cases to test `file -> array`, `array -> array`, and `array -> file`.

**Comment by pseeth** [comment]
Updated the PR in response to comments! And it seems coverage checks are now passing. :)

**Comment by pseeth** [comment]
I think this will require a version bump and associated text in the changelog. Let me know what to do there, or if you'd like, you can add it yourself once this PR is ready to be merged!

**Comment by rabitt** [comment]
I was reviewing and had some comments requiring nontrivial changes so I'm going to push to this branch - 
* create separate functions `_input_format_args(...)` and `_output_format_args(...)` called by `set_input/output_format` and by `build`. 
* rework build to have `input_filepath` and `input_array` as separate inputs
* fix some problems all of this causes with a few special Transformer functions like `.stat`

working on it now!

**Comment by rabitt** [comment]
@pseeth just opened a PR on your fork with those changes.

**Comment by pseeth** [comment]
Looks good to me! Thanks! I merged it into my PR, so now we wait on tests...

**Comment by rabitt** [comment]
I updated the checks so that the docs get built and to let coverage decrease a bit without failing. 

Looks like the docs build is failing because we added numpy, and there are a few lines I added that aren't tested. Going to make one more PR to your PR to fix!

Could you add the changelog notes and maybe add an example to the docs for how to use the new functionality?

**Comment by pseeth** [comment]
Sounds good! I'll update the changelog.

**Comment by pseeth** [comment]
Actually, some guidance on this would be helpful. What should the version bump to?

**Comment by rabitt** [comment]
I found some issues with the docs build (the returns aren't properly documented so they're not showing up in the docs) - I'm going ahead and merging this without bumping the version.

I'll fix the docs and do the version bump in a separate PR.

Thanks!
